File: Data Structures & Algorithms/is-palindrome/submission-1.java
class Solution {

    public boolean isPalindrome(String s) {
        int left = 0, right = s.length() - 1;
        while (left <= right) {
            while (left < right && !isAlphaNumeric(s.charAt(left))) {
                left++;
            }
            while (right > left && !isAlphaNumeric(s.charAt(right))) {
                right--;
            }
            if (Character.toLowerCase(s.charAt(left)) != Character.toLowerCase(s.charAt(right))) return false;

            left++;
            right--;
        }
        return true;
    }

    public static boolean isAlphaNumeric(char c) {
        char ch = Character.toLowerCase(c);
        return (ch >= 'a' && ch <= 'z') || (ch >= '0' && ch <= '9');
    }
    
}

File: Data Structures & Algorithms/sqrtx/submission-1.py
class Solution:
    def mySqrt(self, x: int) -> int:
        if x < 2:
            return x

        left, right = 1, x // 2

        while left <= right:
            mid = (left + right) // 2
            val = mid * mid

            if val == x:
                return mid
            elif val < x:
                left = mid + 1
            else:
                right = mid - 1

        return right